### PR TITLE
ADEN-9996 TCF v2.0 implementation

### PR DESCRIPTION
**Pull request by dmnsobczak**

### CODE REVIEW

File: src/shared/GeoManager.js [hunk 1]
```diff
@@ -88,10 +90,36 @@ function getGeoDataFromCookie(type = 'country') {
...
 class GeoManager {
+    tcf1Disabled = false;
+    tcf2Enabled = false;
+
     constructor(country, region, countriesRequiringPrompt) {
         this.geosRequiringPrompt = (countriesRequiringPrompt || COUNTRIES_REQUIRING_PROMPT).map(country => country.toLowerCase());
         this.country = (country || getGeoDataFromCookie('country') || MISSING_COOKIE_NAME).toLowerCase();
         this.region = (region || getGeoDataFromCookie('region') || MISSING_COOKIE_NAME).toLowerCase();
+
+        this.callInstantConfig().then(() => {console.log(icbmContent);
```
Comment (dmnsobczak) [inline src/shared/GeoManager.js:101]: Remove this debug message.

File: src/shared/GeoManager.js [hunk 2]
```diff
@@ -88,10 +90,36 @@ function getGeoDataFromCookie(type = 'country') {
...
+            this.tcf1Disabled =
+                icbmContent.icTcf1Disabled[0].regions &&
+                icbmContent.icTcf1Disabled[0].regions.includes(this.country.toUpperCase());
+            this.tcf2Enabled =
+                icbmContent.icTcf2Enabled[0].regions &&
+                icbmContent.icTcf2Enabled[0].regions.includes(this.country.toUpperCase());
+
+            console.log('ICBM status', 'tcf1Disabled', this.tcf1Disabled, 'tcf2Enabled', this.tcf2Enabled);
```
Comment (dmnsobczak) [inline src/shared/GeoManager.js:109]: Remove or move to debug mode only

File: src/gdpr/ConsentManagementProvider.js [hunk 3]
```diff
@@ -1,379 +1,214 @@
...
 });
+const debug = (...args) => {
+    const debugQueryParam = 'tracking-opt-in-debug';
+    const isDebug = window.location.search.indexOf(`${debugQueryParam}=true`) !== -1;
 
-function createConsent(consentString, vendorList, options = {}) {
-    if (consentString && !vendorList) {
-        return new ConsentString(consentString);
+    if (isDebug) {
+        console.log('[DEBUG] GDPR: ', ...args);
```
Comment (jacekbj) [inline src/gdpr/ConsentManagementProvider.js:35]: You can always use https://developer.mozilla.org/en-US/docs/Web/API/Console/debug level:debug for that.

File: src/gdpr/ConsentManagementProvider.js [hunk 4]
```diff
@@ -1,379 +1,214 @@
...
-                callback(queue, true);
-            } else {
-                queue.push([commandName, parameter, callback]);
-            }
-        };
+    configure(options) {
+        Object.assign(this.options, options);
 
-        ConsentManagementProvider.addLocatorFrame();
+        debug('Configured with params', options);
```
Comment (jacekbj) [inline src/gdpr/ConsentManagementProvider.js:70]: I'd also log the options passed to constructor.

File: src/gdpr/ConsentManagementProviderLegacy.js [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+        });
+    }
+
+    return map;
+}
+
+class ConsentManagementProviderLegacy {
+    mounted = false;
+
+    static installStub(gdprAppliesGlobally = false) {
```
Comment (jacekbj) [inline src/gdpr/ConsentManagementProviderLegacy.js:68]: There is a static method `installStub` and an instance method. Please add some comments to make it easier to grasp how they're different.

File: src/gdpr/ConsentManagementProviderLegacy.js [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+        return this.fetchVendorList(vendorListVersion)
+            .then((vendorList) => {
+                this.vendorList = vendorList;
+                this.createConsents();
+                this.mount();
+            });
+    }
+
+    installStub(...args) {
+        return ConsentManagementProviderLegacy.installStub(...args);
```
Comment (jacekbj) [inline src/gdpr/ConsentManagementProviderLegacy.js:282]: `ConsentManagementProviderLegacy.installStub` does not accept any parameters.

File: src/shared/GeoManager.js [hunk 7]
```diff
@@ -1,7 +1,9 @@
 import Cookies from 'js-cookie';
+import { getJSON } from './utils';
 
 export const COUNTRY_COOKIE_NAME = 'Geo';
 const MISSING_COOKIE_NAME = 'no-cookie';
+let icbmContent = null;
```
[the remark this answers is not in the document]
Comment (dmnsobczak) [reply src/shared/GeoManager.js:6]: There is many GeoManager instances and I don't want to call ICBM in every of them - they should share this variable.

File: src/shared/GeoManager.js [hunk 8]
```diff
@@ -88,10 +90,36 @@ function getGeoDataFromCookie(type = 'country') {
...
+    tcf2Enabled = false;
+
     constructor(country, region, countriesRequiringPrompt) {
         this.geosRequiringPrompt = (countriesRequiringPrompt || COUNTRIES_REQUIRING_PROMPT).map(country => country.toLowerCase());
         this.country = (country || getGeoDataFromCookie('country') || MISSING_COOKIE_NAME).toLowerCase();
         this.region = (region || getGeoDataFromCookie('region') || MISSING_COOKIE_NAME).toLowerCase();
+
+        this.callInstantConfig().then(() => {console.log(icbmContent);
+            this.tcf1Disabled =
+                icbmContent.icTcf1Disabled[0].regions &&
```
[the remark this answers is not in the document]
Comment (dmnsobczak) [reply src/shared/GeoManager.js:103]: It's not 100% fault-proof, so we need to keep ICBM variables correct (their going to be AdEng level only). I didn't want to reimplement whole InstantConfig service here.

File: src/shared/GeoManager.js [hunk 9]
```diff
@@ -88,10 +90,36 @@ function getGeoDataFromCookie(type = 'country') {
...
+                icbmContent.icTcf1Disabled[0].regions.includes(this.country.toUpperCase());
```
[the remark this answers is not in the document]
Comment (dmnsobczak) [reply src/shared/GeoManager.js:104]: We don't need it. We only need a way to change the status of TCF rollout without making a release.

File: src/shared/utils.js [hunk 10]
```diff
@@ -4,8 +4,8 @@ export const PURPOSES = {
     INFORMATION: 1,
     PERSONALIZATION: 2,
     AD: 3,
-    CONTENT: 3,
-    MEASUREMENT: 3,
+    CONTENT: 4,
```
[the remark this answers is not in the document]
Comment (dmnsobczak) [reply src/shared/utils.js:7]: Based on this:
https://github.com/Wikia/tracking-opt-in/blob/d59e76f7f483bfdd0b72d00c721edff05ca26463/src/components/PreferencesSection.js#L34
I assumed it's a typo - so I fixed it.

File: src/shared/GeoManager.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (dmnsobczak) [reply src/shared/GeoManager.js:103]: Note: add XX support and checking variable value.

File: src/ccpa/UserSignalMechanism.js [hunk 11]
```diff
@@ -27,6 +27,14 @@ const getUSPValue = (value) => {
 
     return value ? USP_VALUES.yes : USP_VALUES.no;
 };
+const debug = (...args) => {
+    const debugQueryParam = 'tracking-opt-in-debug';
```
[the remark this answers is not in the document]
Comment (dmnsobczak) [reply src/ccpa/UserSignalMechanism.js:31]: Done.

File: src/ccpa/UserSignalMechanism.js [hunk 12]
```diff
@@ -27,6 +27,14 @@ const getUSPValue = (value) => {
...
+    const isDebug = window.location.search.indexOf(`${debugQueryParam}=true`) !== -1;
```
[the remark this answers is not in the document]
Comment (dmnsobczak) [reply src/ccpa/UserSignalMechanism.js:32]: Oh, nice! Used. Actually I moved whole debug method to utils.

File: src/modal/Preferences.js [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+        features: null,
+        specialFeatures: null,
+    };
+
+    // ToDo: Rewrite the logic and get rid of this method
+    objectToArray(object) {
+        const array = [];
+
+        for (let [key, value] of Object.entries(object)) {
+            array[key] = value;
```
[the remark this answers is not in the document]
Comment (dmnsobczak) [reply src/modal/Preferences.js:25]: For purposes: yes - it's 1...n
For vendors: not really, there are some blank spots.

File: src/modal/PreferencesVendorList.js [hunk 14]
```diff
@@ -0,0 +1,208 @@
...
+
+    isConsentedPurpose(purposeId) {
+        return this.props.consentedPurposes.indexOf(purposeId) >= 0;
+    }
+
+    isConsentedVendor(vendorId) {
+        return this.props.consentedVendors.indexOf(vendorId) >= 0;
+    }
+
+    getPurposeById(purposeId) {
```
[the remark this answers is not in the document]
Comment (dmnsobczak) [reply src/modal/PreferencesVendorList.js:23]: Removed.

File: src/modal/PreferencesVendorList.js [hunk 15]
```diff
@@ -0,0 +1,208 @@
...
+        return this.props.allPurposes.find(purpose => purpose.id === purposeId);
+    }
+
+    getFeatureById(featureId) {
```
[the remark this answers is not in the document]
Comment (dmnsobczak) [reply src/modal/PreferencesVendorList.js:27]: Removed.

File: src/shared/GeoManager.js (hunk 1 above)
Comment (dmnsobczak) [reply src/shared/GeoManager.js:101]: Removed and added.

[the remark this answers is not in the document]
Comment (dmnsobczak) [reply src/shared/GeoManager.js:101]: Reject callback added

File: src/shared/GeoManager.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (dmnsobczak) [reply src/shared/GeoManager.js:103]: Refactor done.

File: src/shared/GeoManager.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (dmnsobczak) [reply src/shared/GeoManager.js:104]: Value checking added in refactor.

File: src/shared/GeoManager.js (hunk 2 above)
Comment (dmnsobczak) [reply src/shared/GeoManager.js:109]: Done.

File: src/shared/GeoManager.js [hunk 16]
```diff
@@ -88,10 +90,36 @@ function getGeoDataFromCookie(type = 'country') {
...
+            this.tcf2Enabled =
+                icbmContent.icTcf2Enabled[0].regions &&
+                icbmContent.icTcf2Enabled[0].regions.includes(this.country.toUpperCase());
+
+            console.log('ICBM status', 'tcf1Disabled', this.tcf1Disabled, 'tcf2Enabled', this.tcf2Enabled);
+        });
+    }
+
+    callInstantConfig() {
```
[the remark this answers is not in the document]
Comment (dmnsobczak) [reply src/shared/GeoManager.js:113]: Done!

File: src/shared/GeoManager.js [hunk 17]
```diff
@@ -88,10 +90,36 @@ function getGeoDataFromCookie(type = 'country') {
...
+        if (!icbmContent) {
+            icbmContent = getJSON('https://services.wikia.com/icbm/api/config?app=oasis').then((content) => {
```
[the remark this answers is not in the document]
Comment (dmnsobczak) [reply src/shared/GeoManager.js:115]: Added.

File: src/shared/GeoManager.js [hunk 18]
```diff
@@ -88,10 +90,36 @@ function getGeoDataFromCookie(type = 'country') {
...
+                icbmContent = content;
+            });
```
[the remark this answers is not in the document]
Comment (dmnsobczak) [reply src/shared/GeoManager.js:117]: Reject callback added.

File: src/shared/utils.js [hunk 19]
```diff
@@ -4,8 +4,8 @@ export const PURPOSES = {
...
+    MEASUREMENT: 5,
 };
 
 export const FEATURES = {
```
[the remark this answers is not in the document]
Comment (dmnsobczak) [reply src/shared/utils.js:11]: Removed.

File: src/gdpr/ConsentManagementProvider.js (hunk 4 above)
Comment (dmnsobczak) [reply src/gdpr/ConsentManagementProvider.js:70]: Added.

File: src/gdpr/ConsentManagementProvider.js [hunk 20]
```diff
@@ -1,379 +1,214 @@
...
     }
 
-    static addLocatorFrame() {
-        const name = '__cmpLocator';
-
-        if (window.frames[name]) {
-            return;
-        }
+    installStub(...args) {
+        return ConsentManagementProvider.installStub(...args);
```
[the remark this answers is not in the document]
Comment (dmnsobczak) [reply src/gdpr/ConsentManagementProvider.js:74]: Removed.

File: src/gdpr/ConsentManagementProvider.js [hunk 21]
```diff
@@ -1,379 +1,214 @@
...
-    getVendorConsentCookie() {
-        return Cookies.get(VENDOR_CONSENT_COOKIE_NAME) || '';
+        this.mounted = false;
     }
 
-    setVendorConsentCookie(consentString) {
-        const cookieAttributes = this.options.cookieAttributes;
+    updateApi(event) {
+        switch (event) {
+            case 'ui-visible-new':
```
[the remark this answers is not in the document]
Comment (dmnsobczak) [reply src/gdpr/ConsentManagementProvider.js:117]: Done!

File: src/gdpr/ConsentManagementProvider.js [hunk 22]
```diff
@@ -1,379 +1,214 @@
...
+                debug('UI displayed for the first time');
+                break;
 
-        if (consentString) {
-            Cookies.set(VENDOR_CONSENT_COOKIE_NAME, consentString, cookieAttributes);
-        } else {
-            Cookies.remove(VENDOR_CONSENT_COOKIE_NAME, cookieAttributes);
-        }
-    }
+            case 'ui-visible-reset':
```
[the remark this answers is not in the document]
Comment (dmnsobczak) [reply src/gdpr/ConsentManagementProvider.js:122]: Done!

File: src/gdpr/ConsentManagementProvider.js [hunk 23]
```diff
@@ -1,379 +1,214 @@
...
+                debug('UI displayed after policy change');
+                break;
 
-    getPublisherConsentCookie() {
-        return Cookies.get(PUBLISHER_CONSENT_COOKIE_NAME) || '';
-    }
-
-    setPublisherConsentCookie(consentString) {
-        const cookieAttributes = this.options.cookieAttributes;
+            case 'disable':
```
[the remark this answers is not in the document]
Comment (dmnsobczak) [reply src/gdpr/ConsentManagementProvider.js:127]: Done!

File: src/gdpr/ConsentManagementProvider.js [hunk 24]
```diff
@@ -1,379 +1,214 @@
...
 
-        this.setVendorConsentCookie(this.getVendorConsentCookie() || this.vendorConsent.getConsentString());
-    }
+        if (!gdprApplies) {
+            this.cmpApi.update(null);
 
-    isCmpCommand(commandName) {
-        return this.cmpCommands.indexOf(this[commandName]) !== -1;
-    }
+            debug('Not applies');
```
[the remark this answers is not in the document]
Comment (dmnsobczak) [reply src/gdpr/ConsentManagementProvider.js:147]: Changed!

File: src/gdpr/ConsentManagementProvider.js [hunk 25]
```diff
@@ -1,379 +1,214 @@
...
     disableConsentQueue: false,
     gdprApplies: false,
-    gdprAppliesGlobally: false,
-    hasGlobalScope: false,
     language: CMP_DEFAULT_LANGUAGE,
-    vendorList: null
 });
+const debug = (...args) => {
+    const debugQueryParam = 'tracking-opt-in-debug';
+    const isDebug = window.location.search.indexOf(`${debugQueryParam}=true`) !== -1;
```
[the remark this answers is not in the document]
Comment (dmnsobczak) [reply src/gdpr/ConsentManagementProvider.js:32]: Done!

File: src/gdpr/ConsentManagementProvider.js [hunk 26]
```diff
@@ -1,379 +1,214 @@
...
 
-            try {
-                const data = (typeof event.data === 'string') ? JSON.parse(event.data) : event.data;
-                call = data.__cmpCall;
-            } catch (error) {
-                void(error);
-            }
+            this.loaded = ConsentManagementProvider.fetchVendorList()
+                .then((vendorListContent) => {
+                    this.vendorList = vendorListContent;
```
[the remark this answers is not in the document]
Comment (dmnsobczak) [reply src/gdpr/ConsentManagementProvider.js:89]: Done.

**Pull request merged**